cmd/run.go

```diff
@@ -36,87 +35,60 @@ func newCommandSSMRun() *cobra.Command {
+
 	// Get the number of cores available for parallelization
 	runtime.GOMAXPROCS(runtime.NumCPU())
 
-	if len(instanceList) > 0 && len(filterList) > 0 {
-		cmdutil.UsageError(cmd, "The --filter and --instance flags cannot be used simultaneously.")
-		os.Exit(1)
-	}
-
-	if len(instanceList) > 50 {
-		cmdutil.UsageError(cmd, "The --instance flag can only be used to specify a maximum of 50 instances.")
-	}
-
 	// If the --commands and --file options are specified, we append the script contents to the specified commands
 	if inputFile := cmdutil.GetFlagString(cmd, "file"); inputFile != "" {
-		// Open our file for reading
-		file, err := os.Open(inputFile)
-		if err != nil {
-			log.Fatalf("Could not open file at %s\n%s", inputFile, err)
-		}
-
-		defer file.Close()
-
-		// Grab each line of the script and append it to the command slice
-		// Scripts using a line continuation character (\) will work fine here too!
-		scanner := bufio.NewScanner(file)
-		for scanner.Scan() {
-			commandList = append(commandList, scanner.Text())
-		}
-
-		if err := scanner.Err(); err != nil {
-			log.Fatalf("Issue when trying to read input file\n%s", err)
+		if err := util.ReadScriptFile(inputFile, &commandList); err != nil {
+			log.Fatal(err)
 		}
 	}
 
-	if commandList == nil || len(commandList) == 0 {
+	switch { // These cases are all fatal for our invocations or result in undefined behavior
+	case len(instanceList) > 0 && len(filterList) > 0:
+		cmdutil.UsageError(cmd, "The --filter and --instance flags cannot be used simultaneously.")
+	case len(instanceList) > 50:
+		cmdutil.UsageError(cmd, "The --instance flag can only be used to specify a maximum of 50 instances.")
+	case len(commandList) == 0:
 		cmdutil.UsageError(cmd, "Please specify a command to be run on your instances.")
-		os.Exit(1)
+	case len(profileList) > 0 && allProfilesFlag:
+		cmdutil.UsageError(cmd, "The --profile and --all-profiles flags cannot be used simultaneously.")
 	}
 
-	log.Info("Command(s) to be executed: ", strings.Join(commandList, ","))
-
-	if len(profileList) == 0 {
-		env, exists := os.LookupEnv("AWS_PROFILE")
-		if exists {
-			profileList = []string{env}
-		} else {
-			profileList = []string{"default"}
-		}
-	}
-
-	if len(regionList) == 0 {
-		env, exists := os.LookupEnv("AWS_REGION")
-		if exists == false {
-			regionList = []string{env}
-		}
-	}
-
-	// If --all-profiles is set, we call getAWSProfiles() and iterate through the user's ~/.aws/config
-	if allProfilesFlag := cmdutil.GetFlagBool(cmd, "all-profiles"); allProfilesFlag {
-		profileList, err := awsx.GetAWSProfiles()
-		if profileList == nil || err != nil {
-			log.Error("Could not load profiles.", err)
-			os.Exit(1)
-		}
-	}
-
-	// Set up our AWS session for each permutation of profile + region
-	sessionPool := session.NewPoolSafe(profileList, regionList)
-
-	// Convert the filter slice to a map
 	targets := []*ssm.Target{}
 
-	if len(filterList) > 0 {
-		targets = util.SliceToTargets(filterList)
+args:
```

Comment:
This is certainly not as much code but this is throwing up some serious red flags for me. Error handling in go can certainly be verbose but these recent changes are harder to read then what existed before.

Reply:
Propose an alternative?

Reply:
If it's just "go back to using a bunch of nested if blocks", that's fine, but I'd love it if there was a cleaner way to do this.

Reply:
We have two different things within this function that are bloating the size of the function due to tedious checks/error handling. We should extract the validation checks (line 57) to its own function. Focus on single if statements without nesting if possible. The assignment of profileList, targets, and regionList is a little more involved. But the for switch statement is hard to read at first. None of these seem tied to each other so we should handle them separately. If we want to increase readability of the runCommand function its self I think we can extract the logic in getting those values into 3 distinct fucntions. So add something like this:

```go
func validateFlags(instanceList []string, commandList []string, filterList []string) error {
        // Check and return error if parameters are invalid, nil otherwise

        return nil
}

func getProfileList(profileList []string, allProfilesFlag bool) []string, error {
}

func getTargets(filterList []string, targets []*ssm.Target) []*ssm.Target {
}

func getRegionList(regionList []string) []string {
}
```

Then for assignment we would just do 

```go
profileList, err := getProfileList(profileList, allProfilesFlag)
if err {
        log.Fatalf("Could not load profiles.\n%v", err)
}
targets = getFilterList(filterList, targets)
regionList = getRegionList(regionList)
```

Reply:
Thank you for the suggestion, I'll work on adding that. :)

Reply:
Something I noticed about the profiles. If allProfiles is specified we load the list of profiles from the users config. However, if the list of profiles or a specific profile is passed it does not appear we validate that that profile exists within the users configured profile. I think this would be handled within the SafePool session code, but not sure how clear of an error that would be to users. While we are getting the list of profiles should we always perform that lookup and validate the list provided is a valid set of profiles?

Reply:
Great question! We wrap our session creation in [session.Must()](https://github.com/aws/aws-sdk-go/blob/v1.33.9/aws/session/session.go#L324), which is an AWS helper function that just panics if there's an error during creation.

Reply:
Through further research, we discovered that this approach... doesn't actually work. There's no way to know if a session is actually valid until you try and use it. I've added a validation check by calling `session.Config.Credentials.Get()` before we start invocations, along with a small change to our HTTP client timeout to handle testing outside of EC2.